### DISCUSSION TITLE
How can I make this to work with Tiny AVR board and ATTiny85?

### DISCUSSION THREAD

**Discussion by Pierretherenard** [SpenceKonde/ATTinyCore#578] [General]
I use the sparkfun Tiny AVR board.
I plug the board in the USB and even try with an extension USB for either burn bootloader  or upload a basic sketch but I always get:
> Device search timed out!

The only time it was a success is when I did Attiny85(No boot loader)
I am using Arduino 1.8.15 on Mac Pro late 11 OS sierra.
The blink example is working with Damellis

![Screen Shot 2021-06-08 at 6 44 40 PM](https://user-images.githubusercontent.com/23002446/121279687-ccfadf00-c889-11eb-8495-ab2c7c3d3c6b.png)

**Comment by Pierretherenard** [comment]
Was this written before? **Current strongly recommended IDE version: 1.8.13**
**YES it was.**
I was thinking, is it possible that there is so much difference between 1.8.13 and 1.8.15?
 There is. As it is written, I(nobody and everybody), as well strongly recommend 1.8.13.

![Screen Shot 2021-06-09 at 10 24 42 AM](https://user-images.githubusercontent.com/23002446/121401668-d88fe980-c90d-11eb-9e09-0b3d298e96b8.png)

Arduino is working on 2.0.0. I will not update due to my old laptop that will not support this next new version.

Thank you

**Comment by SpenceKonde** [comment]
I am confused as to what you are trying to do and I have no idea what this board you're talking about is. Could you link me to some information? 

If this is a board where you're supposed to program directly over usb (I know there;s one that emulates a USBTinyISP), if you say you used an external programmer to put a sketch on it, the act of using an external isp programmer would have erased the usb bootl;oader. you'd need to find the firmware and manually reload it with isp programmer and avrdude.

**Comment by SpenceKonde** [comment]
this is rthe first I have heard about arduino 1.8.15, just downloaded it now, I hope they haven't done something terminally stupid with programmer menus.

**Comment by SpenceKonde** [comment]
Oh - if it's this thing: https://www.sparkfun.com/products/11801 then that has nothing to do with digispark/micronucleus, you want the ATtiny25/45/85 option and the USBTinyISP slow option for first time programming it, or if you had it set for 1 MHz previously, or the other option when the chip is running fast enough to handle it (2 MHz+ because the "slow" speed is... well.... really slow.

**Reply by Pierretherenard** [reply]
It worked with Arduino 1.8.13 as you are recommending. Yes, this is the product(11801) I am using but on the 1.8.15 it does not work and as you see on the pictures above there are different options compared with the option of 1.8.13.

Thank you for your reply.

I am now on chasing why #include <Wire.h> give me: error: 'TWEN' undeclared  and bunch of others undeclared silliness but I will troubleshoot and read more first.

**Reply by SpenceKonde** [reply]
Okay NOW it all comes into focus. You are using a different core (maybe thje david mellis core now. His core does not include the same programmer definitions, does not have many of the configuration options that mine does, and does not include a special version of Wire.h that suipports the ATtiny85. 

The I2C hardware on the tiny85 is _totally_ _different_ than on the ATmega line, and until the introduction of my Universal Wire library (the Wire.h included with the core), you had to use a special llibrary, like TinyWire. And even if it provided an identical interface, if you wanted to use a library that included Wire.h normally, you had to go modify it to ask for TinyWire. Or whichever of the half dozen of so similar libraries thart were making the rounds. 

A bit more on the I2C hardware on the classic tinyAVRs: with a few exceptions (SPI on 841/441/828/1634. and both on 88), there is no TWI/SPI functionality in hardware on the tinies. Instead they combine the two into a "universal serial interface", a rather lousy excuse for a peripheral that is barely better than bitbanging.  Someone PR'ed the unification of Wire and TinyWire iirc, so I'd never really looked and the very low levels (which, apparently, are just Atmel reference code with tweaks) until a few months ago when someone ran into an issue on the 861 that was specific to that part. I never realized what a  crappy peripheral it was before then; the man who wrote that library has more patience to bad peripherals than I do. But then, he would, since he worked at a microcontroller company back when bad peripherals didn't look quite so ugly. When you're migrating from a part with only terrible peripherals, ones that are just bad look pretty nice. 

The current generation parts (ie, stuff like the ATtiny1614,, 3216, - stuff with a 0, 1, or 2 in the 10's digit and 2. 4, 6. or 7 in the 1's digit) have taken a huge leap forward though = they all have full SPI/TWI/Serial, and a general suite of features that makes an atmega328p jealous, and the same versions of all the periupherals as the full size parts.

**Reply by Pierretherenard** [reply]
Hum, I am using your core and just to make sure I just burned a boot loader. 
How do I make sure I use your core? Is there a signature somewhere?
![Screen Shot 2021-06-09 at 3 17 44 PM](https://user-images.githubusercontent.com/23002446/121437881-d5a9ee80-c937-11eb-9196-9f3558dcb952.png)

**Comment by Pierretherenard** [comment]
Interesting if I use any options in Board-->ATTinyCore-->xxx then I get:> Device search timed out!
but if I don't choose anything and leave as is then it works.

**Reply by SpenceKonde** [reply]
If you use any board from any menu that is not a suybmenu of Board -> ATTinyCore, you're using some other core. 

Please show me full error message of "device search timed out"? I have no idea what that error is.

**Reply by Pierretherenard** [reply]
OK Let me go back to the lastest version since I Tried 3 version of the Arduino IDE. It does not show me any error by the way.
Wierd how come I can use another core after a successful bootloader?

No time tonight. Tomorrow I will show the pictures of what I am doing.

Thank you very much for your time
Pierre

**Reply by pcfreak1201** [reply]
Moin!
Just my 2 cents: "device search timed out" - could it come from a Bluetooth handler? Because the device is not a /dev/***ttyS***

**Reply by SpenceKonde** [reply]
See my amswer below. he was doing "upgrade" mode bootloading for micronucleus.

**Comment by SpenceKonde** [comment]
aaah!!! I figured out what the error message is coming from! That error happens if you select a Digispark/Micronucleus board (one of the boards with direct USB for upload, with the USB handled on the microcontroller itself - which is not what you have - and have Burn Bootloader Method set to upgrade, when no digispark/micronucleus board is connected to a usb port during that period of time. 

If you had a digispark board and plugged it in then, it would use the existing bootloader to upload a new bootloader image and an updater program (as if it were a sketch), that update would then delete the old bootloader and copy the new one into it's place, and then delete itself. Original digispark boards shipped with reset disabled, and some people choose to disable reset on the current-produiction clones to get an extra pin, so that's the only method they have to update the bootloader (which over the years has shrunk from over 3k to under 2k, so there's a lot of reason for those individuals to want to upgrade it), short of using an HVSP programmer, which is a rather exotic tool (as in, not only have I never used one, I don't even have one; at least HV UPDI, I have several takes on it, I just haven't used them.

**Reply by Pierretherenard** [reply]
Ok I think I did it! Let me know if this make sense. I burn bootloader with this options, see below picture.
Then when I upload a basic sketch with the first line #include <Wire.h> there is no more error and the sketch run.
I don't understand why using option 'NO bootloader' actually burn a bootloader.

![Screen Shot 2021-06-09 at 8 48 47 PM](https://user-images.githubusercontent.com/23002446/121461988-b4abc280-c964-11eb-9e53-b489283b7cf7.png)

**Reply by SpenceKonde** [reply]
> I don't understand why using option 'NO bootloader' actually burn a bootloader.

Because if the Arduino folks were being accurate, they would not have named the command "burn bootloader" 

It does two or three things
1. It erases the entire flash memory by executing a Chip Erase cycle: This clears the flash, unless EESAVE is set, it clears the EEPROM too, and then ( and only when those two have been erased), if the chip was "locked" to prevent reading out the flash contents, those lockbits are clearned. 
2. It writes a value (calculated from tools submeus) to the configuration bytes (called "fuses" even though they reprogrammable (I was told by Microchip Bob that they have same write endurance as the main flash) to configure clock source, whether EEPROM is preserved through chip erase, if a bootloader is to be used, and a few other parameters. These configuration bytes, like the device signature, is in a separate address space, and reading them from the application is actually somewhat tricky because while avr/boot.h provides a macro for this.... the parameter it needs is missing from the headers for a great many parts, and known by two different names on the ones where it's present... and while it's usually 5, it isn't always.  (on classic AVRs - on modern AVRs fuses and sigrow are always memory mapped like they always should have been and it's trivial to access them from the application code,)
3. It writes a bootloader image, if any, to the flash (On ATTinyCore 1.5.x, dummy files are written with no bootloader. In 2.0.0 I'll use the solution that megaTinyCore and DxCore use, and lose the dummy file. 

But yeah, they call that burn bootloader. I would love to have a way to override it from platform.txt - but Arduino has not seen fit to provide such a mechanism, nor changed it to say something more informative.
"Configure chip and write bootloader"? "set fuses and burn bootloader" ? 

(note also that on modern AVRs - at least when using my cores - any fuses controlled by menu settings that *can't* leave the chip in a difficult to reprogram state get set whenever uploading via UPDI (the modern AVR version of ISP). But on classic AVRs, the three fusebytes, generally, all have at least one bit in them that can softbrick the chip. Stuff like that is best left for "burn bootloader" since people won't throw that around like they would upload.

**Comment by Pierretherenard** [comment]
I just used the above options to burn bootloader using 8Mhz on a TinyAVR board.
This is working and i just want to show the information I receive from the process.

Thank you

For information, here is the result of the bootloader using the above options:

avrdude: Version 6.3-20201216
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch

         System wide configuration file is "/Users/xxx/Library/Arduino15/packages/ATTinyCore/hardware/avr/1.5.2/avrdude.conf"
         User configuration file is "/Users/xxx/.avrduderc"
         User configuration file does not exist or is not a regular file, skipping

         Using Port                    : usb
         Using Programmer              : usbtiny
         Setting bit clk period        : 0.3
avrdude: usbdev_open(): Found USBtinyISP, bus:device: 250:005
         AVR Part                      : ATtiny85
         Chip Erase delay              : 400000 us
         PAGEL                         : P00
         BS2                           : P00
         RESET disposition             : possible i/o
         RETRY pulse                   : SCK
         serial program mode           : yes
         parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         ByteDelay                     : 0
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                  Block Poll               Page                       Polled
           Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom        65    12     4    0 no        512    4      0  4000  4500 0xff 0xff
           flash         65    12    32    0 yes      8192   64    128 30000 30000 0xff 0xff
           signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00
           lock           0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           lfuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           hfuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           efuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           calibration    0     0     0    0 no          2    0      0     0     0 0x00 0x00

         Programmer Type : USBtiny
         Description     : USBtiny simple USB programmer, http://www.ladyada.net/make/usbtinyisp/
avrdude: programmer operation not supported

avrdude: Setting SCK period to 1 usec
avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e930b (probably t85)
avrdude: erasing chip
avrdude: Setting SCK period to 1 usec
avrdude: reading input file "0xFF"
avrdude: writing efuse (1 bytes):

Writing | ################################################## | 100% 0.00s

avrdude: 1 bytes of efuse written
avrdude: verifying efuse memory against 0xFF:
avrdude: load data efuse data from input file 0xFF:
avrdude: input file 0xFF contains 1 bytes
avrdude: reading on-chip efuse data:

Reading | ################################################## | 100% 0.00s

avrdude: verifying ...
avrdude: 1 bytes of efuse verified
avrdude: reading input file "0b11010111"
avrdude: writing hfuse (1 bytes):

Writing | ################################################## | 100% 0.00s

avrdude: 1 bytes of hfuse written
avrdude: verifying hfuse memory against 0b11010111:
avrdude: load data hfuse data from input file 0b11010111:
avrdude: input file 0b11010111 contains 1 bytes
avrdude: reading on-chip hfuse data:

Reading | ################################################## | 100% 0.00s

avrdude: verifying ...
avrdude: 1 bytes of hfuse verified
avrdude: reading input file "0xE2"
avrdude: writing lfuse (1 bytes):

Writing | ################################################## | 100% 0.00s

avrdude: 1 bytes of lfuse written
avrdude: verifying lfuse memory against 0xE2:
avrdude: load data lfuse data from input file 0xE2:
avrdude: input file 0xE2 contains 1 bytes
avrdude: reading on-chip lfuse data:

Reading | ################################################## | 100% 0.00s

avrdude: verifying ...
avrdude: 1 bytes of lfuse verified
avrdude: reading input file "/Users/xxx/Library/Arduino15/packages/ATTinyCore/hardware/avr/1.5.2/bootloaders/empty/empty_all.hex"
avrdude: writing flash (2 bytes):

Writing | ################################################## | 100% 0.06s

avrdude: 2 bytes of flash written
avrdude: verifying flash memory against /Users/xxx/Library/Arduino15/packages/ATTinyCore/hardware/avr/1.5.2/bootloaders/empty/empty_all.hex:
avrdude: load data flash data from input file /Users/xxx/Library/Arduino15/packages/ATTinyCore/hardware/avr/1.5.2/bootloaders/empty/empty_all.hex:
avrdude: input file /Users/xxx/Library/Arduino15/packages/ATTinyCore/hardware/avr/1.5.2/bootloaders/empty/empty_all.hex contains 2 bytes
avrdude: reading on-chip flash data:

Reading | ################################################## | 100% 0.05s

avrdude: verifying ...
avrdude: 2 bytes of flash verified

avrdude done.  Thank you.

**Reply by SpenceKonde** [reply]
Glad it's sorted
